src/main/java/org/opentripplanner/middleware/itinerarymatching/LegMatcher.java

```diff
@@ -0,0 +1,156 @@
+        // - The leg has the same interlining qualities with the previous leg
+        // - The legs are scheduled for the same time of the day. A check is being done for the exact
+        //   scheduled time in order for the trip monitor to attempt to track a specific trip.
+        //   Trip IDs may change over time, however, as far as an end-user is concerned, as long as the same route
+        //   comes at the same time to the same start and end stops, it can be considered a match.
+        List<Match> criteria = List.of(
+            new Match(() -> stopsMatch(referenceLeg.from, candidateLeg.from), "Origin stops"),
+            new Match(() -> stopsMatch(referenceLeg.to, candidateLeg.to), "Destination stops"),
+            new Match(() -> equalsOrReferenceWasNull(referenceLeg.mode, candidateLeg.mode), "Modes"),
+            new Match(() -> agenciesMatch(referenceLeg.agency, candidateLeg.agency), "Agency"),
+            new Match(() -> routesMatch(referenceLeg.route, candidateLeg.route), "Routes"),
+            new Match(() -> equalsIgnoreCaseOrReferenceWasEmpty(referenceLeg.headsign, candidateLeg.headsign), "Headsigns"),
+            new Match(() -> referenceLeg.interlineWithPreviousLeg.equals(candidateLeg.interlineWithPreviousLeg), "Prev. leg interline"),
+            new Match(() -> timeOfDayMatches(referenceLeg.getScheduledStartTime(), candidateLeg.getScheduledStartTime()), "Scheduled start time"),
+            new Match(() -> timeOfDayMatches(referenceLeg.getScheduledEndTime(), candidateLeg.getScheduledEndTime()), "Scheduled end time")
+        );
+
+        result = Match.all(criteria);
+        return result.isSuccessful();
+    }
+
+    /**
+     * Checks whether two stops (OTP Places) match for the purposes of matching itineraries
+     */
+    private static boolean stopsMatch(Place stopA, Place stopB) {
+        // Stop names must match. It's possible in OTP to have a null place name, although it probably won't occur with
+        // transit legs. But just in case this method is expanded in scope to check more stuff about a place, if both
+        // are null, then assume a match.
+        if (
+            (stopA.name != null && !stopA.name.equalsIgnoreCase(stopB.name)) ||
+                (stopA.name == null && stopB.name != null)
+        ) {
+            return false;
+        }
+
+        // stop code must match
+        if (
+            stopA.stop != null &&
+            stopB.stop != null &&
+            !equalsIgnoreCaseOrReferenceWasEmpty(stopA.stop.code, stopB.stop.code)
+        ) {
+            return false;
+        }
+
+        // stop positions must be no further than 5 meters apart
+        double stopDistanceMeters = DistanceUtils.radians2Dist(
+            DistanceUtils.distHaversineRAD(stopA.lat, stopA.lon, stopB.lat, stopB.lon),
+            DistanceUtils.EARTH_MEAN_RADIUS_KM
+        ) * 1000;
+        if (stopDistanceMeters > 5) {
```

Comment:
Consider simplifying and merging with the return i.e. `return !(stopDistanceMeters > 5);`